# mHapSuite
mHapSuite is a tool kit for analysis of DNA methylation haplotypes.
It has 12 sub-commands: convert, merge, tanghulu, mHapView, MHBDiscovery, stat, genomeWide, scatterPlot, boxPlot, heatMapPlot, profilePlot and enrichmentPlot.
The source code of mHapSuite is hosted in [github](https://github.com/yoyoong/mHapSuite). It is written in java, and should run in Windows, Linux and macOS with java1.8 installed.

## Prerequisites
The mHapSuite requires the following dependencies:
* java jdk 1.8.0

## Installation
Push the source and run the jar:
```Shell
git clone https://github.com/yoyoong/mHapSuite
cd mHapSuite/target
java -jar mHapSuite-2.0-jar-with-dependencies.jar [convert/merge/tanghulu/mHapView/MHBDiscovery/stat/genomeWide/scatterPlot/boxPlot/heatMapPlot/profilePlot/enrichmentPlot ...]
```

## convert
Convert SAM/BAM format file to mHap format file. 
It takes an indexed Bisulfite-seq BAM and CpGs position files as inputs to extract DNA methylation haplotypes.
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar convert --inputFile <in.bam> --cpgPath <CpG.gz> [--region chr:start-end] [--bedPath <in.bed>] --outputFile <out> [--nonDirectional] [--mode mode] [--pat pat]
Options:
 -inputFile,--inputFile <args>      input file, SAM/BAM format, should be sorted by samtools
 -cpgPath,--cpgPath <args>          genomic CpG file, gz format and indexed
 -region,--region <args>            one region, in the format of chr:start-end
 -bedPath,--bedPath <args>          bed file, one query region per line
 -outPutFile,--outPutFile <args>    output filename, default: out.mhap.gz
 -nonDirectional,--nonDirectional   do not group results by the direction of reads.
 -mode,--mode <args>                sequencing mode. ( TAPS | BS (default))
 -pat,--pat                         whether inputPath is pat file
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar convert -inputFile SRX8472828.bam -cpgPath hg19_CpG.gz -outPutFile test.mhap.gz nonDirectional -mode BS
```

## merge
Merge multiple sorted mHap files, produce a single sorted mHap file.
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar merge --inputFile <1.mhap.gz 2.mhap.gz ...> --cpgPath <CpG.gz> --outputFile <out>
Options:
 -inputFile,--inputFile <args>     multiple input file, SAM/BAM format, should be sorted by samtools
 -cpgPath,--cpgPath <args>         genomic CpG file, gz format and indexed
 -outPutFile,--outPutFile <args>   output filename, default: out.mhap.gz
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar merge -inputFile SRX8208800.mhap.gz SRX8208801.mhap.gz -cpgPath hg19_CpG.gz -outPutFile out.mhap.gz 
```

## tanghulu
Visualize the methylation status of the reads that cover a given region.
Tanghulu plot was originally introduced in [CGmapTools](https://cgmaptools.github.io/).
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar tanghulu --mhapPath <mhap.gz> --cpgPath <CpG.gz> --region chr:start-end --tag <out> [--outFormat outFormat] [--strand strand] [--maxReads maxReads] [--maxLength maxLength] [--merge] [--simulation] [--cutReads]
Options:
 -mhapPath,--mhapPath <arg>       input file, mhap.gz format, generated by mHapTools and indexed
 -cpgPath,--cpgPath <arg>         genomic CpG file, gz format and indexed
 -region,--region <arg>           one region, in the format of chr:start-end
 -tag,--tag <arg>                 prefix of the output file(s)
 -outFormat,--outFormat <arg>     output format,pdf or png [pdf]
 -strand,--strand <arg>           plus,minus,both [both]
 -maxReads,--maxReads <arg>       the max number of reads to plot [50]
 -maxLength,--maxLength <arg>     the max length of region to plot [2000]
 -merge,--merge                   indicates whether identical mHaps should be merged
 -simulation,--simulation         indicates whether mHaps should be simulated
 -cutReads,--cutReads             indicates whether only keep CpGs in the defined region
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar tanghulu -mhapPath esophagus_T.mhap.gz -cpgPath hg19_CpG.gz -region chr1:10469-10903 -tag esophagus_T -strand both -outFormat png -maxReads 50 -maxLength 2000 -merge 
```

## mHapView
mHapView
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar mHapView --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--region chr:start-end] [--bedPath <in.bed>] --tag <tag> [--outFormat outFormat] [--strand strand]
Options:
 -mhapPath,--mhapPath <arg>     input file, mhap.gz format, generated by mHapTools and indexed
 -cpgPath,--cpgPath <arg>       genomic CpG file, gz format and indexed
 -region,--region <arg>         one region, in the format of chr:start-end
 -bedPath,--bedPath <arg>       a bed file
 -tag,--tag <arg>               prefix of the output file(s)
 -outFormat,--outFormat <arg>   output format,pdf or png [pdf]
 -strand,--strand <arg>         plus,minus,both [both]
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar mHapView -mhapPath esophagus_T.mhap.gz -cpgPath hg19_CpG.gz -region chr1:2121159-2121449 -bedPath CRC_sc_bulk.bed -tag esophagus_T -outFormat png -strand both
```

## stat
Calculate the methylation statistics in a single region or multiple regions.
When calculating a single interval, the --region parameter should be defined.
If there are multiple intervals, a BED file needs to be specified.
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar stat --metrics <metrics> --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--region chr:start-end] [--bedPath <in.bed>] --outputFile <out.tsv> [--minK minK] [--maxK maxK] [--K K] [--strand strand] [--cutReads] [--r2Cov r2Cov]
Options:
 -metrics,--metrics <arg>         mHap-level metrics, including MM, PDR, CHALM, MHL, MCR, MBS, Entropy and R2 [None]
 -mhapPath,--mhapPath <arg>       input file, mhap.gz format, generated by mHapTools and indexed
 -cpgPath,--cpgPath <arg>         genomic CpG file, gz format and indexed
 -region,--region <arg>           one region, in the format of chr:start-end
 -bedPath,--bedPath <arg>         input BED file
 -outputFile,--outputFile <arg>   output filename, default: stat.out.tsv
 -minK,--minK <arg>               minimum k-mer length for MHL [1]
 -maxK,--maxK <arg>               maximum k-mer length for MHL [10]
 -K,--K <arg>                     k-mer length for entropy, PDR, and CHALM, can be 3, 4, or 5 [4]
 -cutReads,--cutReads             indicates whether only keep CpGs in the defined region
 -strand,--strand <arg>           plus,minus,both [both]
 -r2Cov,--r2Cov <arg>             minimal number of reads that cover two CpGs for R2 calculation [20]
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar stat -metrics MM PDR CHALM MHL MCR MBS Entropy R2 -mhapPath esophagus_T.mhap.gz -cpgPath hg19_CpG.gz -bedPath CRC_MHB_non_NC.bed -outputFile stat.out.tsv -minK 1 -maxK 10 -K 4 -strand both -r2Cov 20
```
## genomeWide
Calcuate mHap-level metrics for all reads that cover each individual CpG sites in a genome-wide manner and save the results as bedGraph files.
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar genomeWide --metrics <metrics> --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--region chr:start-end] [--bedPath <in.bed>] --outputDir <outputDir> --tag <tag> [--minK minK] [--maxK maxK] [--K K] [--strand strand] [--cpgCov cpgCov] [--r2Cov r2Cov] [--k4Plus k4Plus]
Options:
 -metrics,--metrics <arg>       mHap-level metrics,including Cov, MM, PDR, CHALM, MHL, MCR, MBS, Entropy, and R2
 -mhapPath,--mhapPath <arg>     input file, mhap.gz format, generated by mHapTools and indexed
 -cpgPath,--cpgPath <arg>       genomic CpG file, gz format and indexed
 -region,--region <arg>         one region, in the format of chr:start-end
 -bedPath,--bedPath <arg>       a bed file, one query region per line
 -outputDir,--outputDir <arg>   output directory, created in advance
 -tag,--tag <arg>               prefix of the output file(s), default: genomeWide_out
 -minK,--minK <arg>             minimum k-mer length for MHL [1]
 -maxK,--maxK <arg>             maximum k-mer length for MHL [10]
 -K,--K <arg>                   k-mer length for entropy, PDR, and CHALM, can be 3, 4, or 5 [4]
 -strand,--strand <arg>         plus,minus,both [both]
 -cpgCov,--cpgCov <arg>         minimal number of CpG coverage for MM calculation [5]
 -r2Cov,--r2Cov <arg>           minimal number of reads that cover two CpGs for R2 calculation [20]
 -k4Plus,--k4Plus <arg>         minimal number of reads that cover 4 or more CpGs for PDR, CHALM, MHL, MCR, MBS and Entropy [5]
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar genomeWide -tag test -mhapPath esophagus_T.mhap.gz -cpgPath hg19_CpG.gz -metrics Cov MM PDR CHALM MHL MCR MBS R2 -outputDir genomeWide_outdir -minK 1 -maxK 10 -K 4 -strand both -cpgCov 5 -r2Cov 20 -k4Plus 5 
```

## MHBDiscovery
DNA methylation of adjacent CpG sites can be co-methylated and forms methylation haplotype blocks (MHBs).
It was original introduced in [S. Guo et al., 2017](https://pubmed.ncbi.nlm.nih.gov/28263317/).
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar MHBDiscovery --mhapPath <mhap.gz> --cpgPath <CpG.gz> [--region chr:start-end] [--bedPath <in.bed>] -outputDir outputDir --tag <tag> [--window window] [--r2 r2] [--pvalue pvalue]
Options:
 -mhapPath,--mhapPath <arg>     input file, mhap.gz format, generated by mHapTools and indexed
 -cpgPath,--cpgPath <arg>       genomic CpG file, gz format and indexed
 -region,--region <arg>         one region, in the format of chr:start-end
 -bedPath,--bedPath <arg>       input BED file
 -outputDir,--outputDir <arg>   output directory, created in advance
 -tag,--tag <arg>               prefix of the output file(s) default: MHBDiscovery_out
 -window,--window <arg>         Size of core window
 -r2,--r2 <arg>                 R square cutoff
 -pvalue,--pvalue <arg>         P value cutoff
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar MHBDiscovery -mhapPath esophagus_T.mhap.gz -cpgPath hg19_CpG.gz -bedPath CRC_MHB_non_NC.bed -window 5 -r2 0.5 -pvalue 0.05 -outputDir outputDir -tag CRC_hg19_MHB -qcFlag 
```

## scatterPlot
A scatter plot which the X-axis is one metric and the Y-axis is another metric from the same sample or different samples.
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar scatterPlot --bedPath <in.bed> --bigwig1 <first.bw> --bigwig2 <second.bw> --tag <tag> --outFormat <outFormat> 
Options:
 -bedPath,--bedPath <arg>       a bed file, one query region per line
 -bigwig1,--bigwig1 <arg>       the first input bigwig file of a metrics rom a sample
 -bigwig2,--bigwig2 <arg>       the second input bigwig file of another metrics, from the same sample of bigwig1, or from another sample
 -tag,--tag <arg>               prefix of the output file(s)
 -outFormat,--outFormat <arg>   output format,pdf or png [pdf]
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar scatterPlot -bedPath MHB_cervix_normal.bed -bigwig1 P1_ESCC_MM.bw -bigwig2 P1_BRCA_MM.bw -tag scatterPlot -outFormat png 
```

## boxPlot
A box plot which show the distribution of metric values from multiple samples.
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar boxPlot --bedPath <in.bed> --bigwigs <first.bw second.bw> --tag <tag> --outFormat <outFormat> 
Options:
 -bedPath,--bedPath <arg>       a bed file, one query region per line
 -bigwigs,--bigwigs <arg>       multi bigwig files, split by blank character
 -tag,--tag <arg>               prefix of the output file(s)
 -outFormat,--outFormat <arg>   output format,pdf or png [pdf]
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar boxPlot -bedPath MHB_cervix_normal.bed -bigwigs P1_BRCA_MM.bw P1_CESC_MM.bw P1_CRC_MM.bw P1_ESCC_MM.bw P1_HNSC_MM.bw P1_LIHC_MM.bw P1_LUADLUSC_MM.bw -tag boxPlot.output -outFormat png
```

## heatMapPlot
A hybrid plot include 2 part: the upper plot is line plot show the average metric values of each windows, 
the lower plot is a heatmap plot show the average metric values of each windows in each regions from multi bed files.
You can refer to [plotHeatmap](https://deeptools.readthedocs.io/en/develop/content/tools/plotHeatmap.html).
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar heatMapPlot --bedPaths <first.bed second.bed> --bigwig <in.bw> [--upLength <upLength>] [--downLength <downLength>] [--window <window>] [--sortRegions <sortRegions>] --tag <tag> --outFormat <outFormat> [-matrixFlag]
Options:
 -bedPaths,--bedPaths <arg>       multi bed files path, one query region per line in a bed file
 -bigwig,--bigwig <arg>           the input bigwig file of a metrics from a sample
 -upLength,--upLength <arg>       the length of upstream region from the center point [20000]
 -downLength,--downLength <arg>   the length of downstream region from the center point [20000]
 -window,--window <arg>           the length of the window [1000]
 -sortRegions,--sortRegions <arg> the sort of the region in heatmap, can be keep, descend, ascend and missingValues [missingValues]
 -tag,--tag <arg>                 prefix of the output file(s)
 -outFormat,--outFormat <arg>     output format,pdf or png [pdf]
 -matrixFlag,--outFormat          whether generate matrix file
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar heatMapPlot -bedPaths MHB_cervix_normal.bed esophagus_normal_MHB.bed -upLength 20000 -downLength 20000 -window 1000 -sortRegions keep -bigwig P1_BRCA_MM.bw -tag heatMapPlot.output -outFormat png -matrixFlag
```

## profilePlot
A line plot show the average metric values of each windows in core/upstream/downstream of the regions from multi bed files.
You can refer to [plotProfile](https://deeptools.readthedocs.io/en/develop/content/tools/plotProfile.html).
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar profilePlot --bedPaths <first.bed second.bed> --bigwig <in.bw> [--upLength <upLength>] [--downLength <downLength>] [--windowNum <windowNum>] --tag <tag> --outFormat <outFormat> [-matrixFlag]
Options:
 -bedPaths,--bedPaths <arg>       multi bed files path, one query region per line in a bed file
 -bigwig,--bigwig <arg>           the input bigwig file of a metrics from a sample
 -upLength,--upLength <arg>       the length of upstream region from the center point [2000]
 -downLength,--downLength <arg>   the length of downstream region from the center point [2000]
 -windowNum,--windowNum <arg>     the window number of the core region [10]
 -tag,--tag <arg>                 prefix of the output file(s)
 -outFormat,--outFormat <arg>     output format,pdf or png [pdf]
 -matrixFlag,--outFormat          whether generate matrix file
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar profilePlot -bedPaths MHB_cervix_normal.bed esophagus_normal_MHB.bed hg19_cpgisland.bed -upLength 2000 -downLength 2000 -windowNum 10 -bigwig P1_BRCA_MM.bw -tag profilePlot.output -outFormat png -matrixFlag
```

## enrichmentPlot
A line plot show the perventage of genomic feature overlap with TCGA-ATAC peaks.
```Shell
usage: java -jar mHapSuite-2.0-jar-with-dependencies.jar enrichmentPlot --bedPaths <first.bed second.bed> --openChromatin <openChromatin.bed> --bigwig <in.bw> [--groupNum <groupNum>] [--groupCutoff <groupCutoff>] --tag <tag> --outFormat <outFormat> 
Options:
 -bedPaths,--bedPaths <arg>             multi bed files path, one query region per line in a bed file
 -openChromatin,--openChromatin <arg>   a open chromatin regions file, one region per line
 -bigwig,--bigwig <arg>                 the input bigwig file of a metrics from a sample
 -groupNum,--groupNum <arg>             the group number of the metrics value [10]
 -groupCutoff,--groupCutoff <arg>       the minimum valid value number in the group [100]
 -tag,--tag <arg>                       prefix of the output file(s)
 -outFormat,--outFormat <arg>           output format,pdf or png [pdf]
 ```
Example of usage:
```Shell
java -jar mHapSuite-2.0-jar-with-dependencies.jar enrichmentPlot -bedPaths MHB_cervix_normal.bed esophagus_normal_MHB.bed -openChromatin hg19_cpgisland.bed -groupNum 10 -groupCutoff 100 -bigwig P1_BRCA_MM.bw -tag enrichmentPlot.output -outFormat png 
```